### ISSUE TITLE
Load fixture issues

### ISSUE THREAD

**Issue by skosito** [NomicFoundation/hardhat#3249]
Sometimes using same fixture in lot of tests throws this error:

```
 FixtureSnapshotError: There was an error reverting the snapshot of the fixture.

This might be caused by using nested loadFixture calls in a test, for example by using multiple beforeEach calls. This is not supported yet.
      at loadFixture (node_modules/@nomicfoundation/hardhat-network-helpers/src/loadFixture.ts:47:15)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async recoverViaDistributor (test/Wallet2.ts:111:9)
```

There is no nested loadFixture, tests are working properly until some amount of loadFixture calls for same fixture across multiple files (therefore multiple describe blocks), but after, for example 10-15 usages it starts failing. If I take same fixture and just rename it and start using it, it works. Is there some limit or best practices on how to handle using same fixture across a lot of files and tests?

**Comment by fvictorio** [comment]
@skosito any chance of creating an [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) for this problem? I know that these kinds of issues are hard to reduce to its minimal expression, so the second best thing would be to have access to the repo and to have clear reproduction steps.

**Comment by skosito** [comment]
> @skosito any chance of creating an [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) for this problem? I know that these kinds of issues are hard to reduce to its minimal expression, so the second best thing would be to have access to the repo and to have clear reproduction steps.

thanks for reply, access to repo is not possible unfortunately, i will try to create reproducible example

**Comment by skosito** [comment]
@fvictorio i am sorry for delay, but here is repo  i put up with min reproduction https://github.com/skosito/hardhat-load-fixture-min-repro, to reproduce just run 
```
npm install
npx hardhat test
```

**Comment by fvictorio** [comment]
Thank your for creating the repro @skosito, I'll take a look as soon as I have some time.

**Comment by fvictorio** [comment]
Ugh, I see what's going on now. This seems to be a bug of our current implementation.

Thanks a lot for the reproduction example, it was super useful.

**Comment by skosito** [comment]
> Ugh, I see what's going on now. This seems to be a bug of our current implementation.
> 
> Thanks a lot for the reproduction example, it was super useful.

@fvictorio thank you for checking it out, let me know if there is a way to help more, happy to contribute if possible

**Comment by skosito** [comment]
@fvictorio sorry for another ping, but i just looked at the code and it seems the issue is that i revert from snapshot 2 to previous snapshot 1 (which is fine), but then try to revert to snapshot 2 which is future state, but blockchain is back to previous state - is that the issue or something else?

is solution to this problem to have better documentation and let developers take care of this when writing tests, or to maybe every time revert to snapshot happens, all other snapshots with higher snapshotid are removed from snapshot array, and have to be re-added again? because it doesnt make sense to revert to future state, so next time snapshot is moved from 1 to 2, 2 has to execute takeSnapshot again?

pseudo code:

```
const restoredSnapshotId = await snapshot.restorer.restore();

for (let snapshot of snapshots) {
remove all snapshots with snapshotId > restoredSnapshotId

}

```

or maybe something more clever with evm_increaseTime if possible?

just trying to understand and learn about this a bit - feel free to ignore, and i can check the fix at the end, just curious

**Comment by fvictorio** [comment]
> or to maybe every time revert to snapshot happens, all other snapshots with higher snapshotid are removed from snapshot array, and have to be re-added again?

I agree, I think something like this is the solution. It seems like a lot of people are running into this, which sucks, so we'll try to prioritize it.

**Comment by skosito** [comment]
@fvictorio i tried this fix in project i am working on and it fixed the issues (project has 130 tests and execution time is the same, previously it was failing for some tests because of issue we are discussing here): https://github.com/skosito/hardhat/commit/82a8f327ff0d333cbd99a7a7680db6bd50af5b40

The only hardhat test failing after the change is this one:

```
 it("should throw the right error when an invalid snapshot is reverted", async function () {
    async function mineBlockFixture() {
      await mineBlock();
    }
    async function mineTwoBlocksFixture() {
      await mineBlock();
      await mineBlock();
    }

    await loadFixture(mineBlockFixture);
    await loadFixture(mineTwoBlocksFixture);
    await loadFixture(mineBlockFixture);
    await assert.isRejected(
      loadFixture(mineTwoBlocksFixture),
      FixtureSnapshotError
    );
  });
```

But to me it looks like that is exactly situation from my repro example:
1. load fixture 1 (snapshot created)
2. load fixture 2 (snapshot created)
3. load fixture 1 (snapshot exists, revert)
4. load fixture 2 (invalid because it was reverted to previous one already in step 3)

with fix above, this test passes now because in step 4 when load fixture 2 happens, that snapshot is removed already in step 3, and it will just takeSnapshot again

hope this helps, or maybe even i can open PR?

**Comment by fvictorio** [comment]
Closed by #3327, this will be included in the next release.
